### PR TITLE
api: add `using:` attribute to head & stable URLs

### PR DESCRIPTION
**Pull request by EricFromCanada**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
Some formulae rely on the `using:` attribute to indicate how a resource is to be downloaded if it can't be detected via the domain name or protocol, e.g.:
- [libvorbis](https://formulae.brew.sh/formula/libvorbis)'s stable URL has `using: :homebrew_curl` as it's hosted on a TLS-1.3-only server
- [trino](https://formulae.brew.sh/formula/trino)'s stable URL has `using: :nounzip` to allow for manual hard-link-preserving decompression
- the head URLs for [abcl](https://formulae.brew.sh/formula/abcl), [global](https://formulae.brew.sh/formula/global), [nginx](https://formulae.brew.sh/formula/nginx), [objfw](https://formulae.brew.sh/formula/objfw), [squid](https://formulae.brew.sh/formula/squid) each have a `using` attribute to indicate the VCS needed to fetch their sources

This change allows from-source installations of these and similar formulae to work when using the API.

Current output:
```console
$ brew info --json pytouhou | jq .[0].urls
{
  "stable": {
    "url": "https://hg.linkmauve.fr/touhou",
    "tag": null,
    "revision": "5270c34b4c00",
    "checksum": null
  },
  "head": {
    "url": "https://hg.linkmauve.fr/touhou",
    "branch": null
  }
}

$ brew fetch --HEAD blastem
==> Downloading https://www.retrodev.com/repos/blastem
##################################################################################################################################################################### 100.0%
Downloaded to: /Users/eric/Library/Caches/Homebrew/downloads/ae00afc1c31ad2cdd9ef6e4ea0876edc8206c3c47c11a628a8593d838242aca5--blastem
SHA256: 72b87d3a585b96e977f9f29cf6fcd9dcee0e9026fc3c69d0b9fdb16b35e679e9
# this just downloads a web page
```

Future:
```console
$ HOMEBREW_NO_INSTALL_FROM_API=1 brew info --json pytouhou | jq .[0].urls
{
  "stable": {
    "url": "https://hg.linkmauve.fr/touhou",
    "tag": null,
    "revision": "5270c34b4c00",
    "checksum": null,
    "using": "hg"
  },
  "head": {
    "url": "https://hg.linkmauve.fr/touhou",
    "branch": null,
    "using": "hg"
  }
}

$ HOMEBREW_NO_INSTALL_FROM_API=1 brew fetch --HEAD blastem
==> Cloning https://www.retrodev.com/repos/blastem
requesting all changes
adding changesets
adding manifests
adding file changes
added 2346 changesets with 5011 changes to 324 files (+1 heads)
new changesets 2432d177e1ac:c76c81c21ae5
updating to branch default
300 files updated, 0 files merged, 0 files removed, 0 files unresolved
Resource: vasm
==> Downloading http://phoenix.owl.de/tags/vasm1_8i.tar.gz
##################################################################################################################################################################### 100.0%
Downloaded to: /Users/eric/Library/Caches/Homebrew/downloads/4f31e8fb49321255218db07092e8864e9e036e14c2c563c1add9ea1201164d99--vasm1_8i.tar.gz
SHA256: 9ae0b37bca11cae5cf00e4d47e7225737bdaec4028e4db2a501b4eca7df8639d
==> Downloading https://www.retrodev.com/repos/blastem/raw-rev/dbbf0100f249
##################################################################################################################################################################### 100.0%
Downloaded to: /Users/eric/Library/Caches/Homebrew/downloads/70a17ee109695d03f5ece64dd949f1b08371b9e3592a96d1738aac5868fba8c3--dbbf0100f249
SHA256: e332764bfa08e08e0f9cbbebefe73b88adb99a1e96a77a16a0aeeae827ac72ff
```

### CODE REVIEW

**Comment by Bo98** [comment]
Will need adding here too: https://github.com/Homebrew/brew/blob/0932f361dceb4e621024eed9c5e6bdd0b55126e2/Library/Homebrew/formulary.rb#L244

plus the equivalent a few lines down for head

File: Library/Homebrew/formula.rb [hunk 1]
```diff
@@ -2281,6 +2281,7 @@ def to_hash
         "url"      => stable_spec.url,
         "tag"      => stable_spec.specs[:tag],
         "revision" => stable_spec.specs[:revision],
+        "using"    => stable_spec.using,
```
Comment (Bo98) [inline Library/Homebrew/formula.rb:2284]: `using` can be a `Class` and we're probably not going to ever support that.

```suggestion
        "using"    => (stable_spec.using if stable_spec.using.is_a?(Symbol)),
```

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -241,7 +241,11 @@ def self.load_formula_from_api(name, flags:)
 
       if (urls_stable = json_formula["urls"]["stable"].presence)
         stable do
-          url_spec = { tag: urls_stable["tag"], revision: urls_stable["revision"] }.compact
+          url_spec = {
+            tag:      urls_stable["tag"],
+            revision: urls_stable["revision"],
+            using:    urls_stable["using"],
```
Comment (Bo98) [inline Library/Homebrew/formulary.rb:247]:
```suggestion
            using:    urls_stable["using"].to_sym,
```

**Comment by MikeMcQuaid** [comment]
Thanks @EricFromCanada!

**Pull request merged**